solutions/noisg-19-feast.mdx

```diff
@@ -0,0 +1,83 @@
+---
+id: noisg-19-feast
+source: NOI
+title: 2019 - Feast
+author: Dong Liu
+---
+
+**Time Complexity:** $\mathcal{O}(N\log{\sum a[i]})$.
+
+<LanguageSection>
+
+<CPPSection>
+
+Like a normal Alien Trick problem, we're going to binary search on the maximum $\texttt{penalty}$.
+
+Now, for a given penalty, we're going to calculate the number of people in an optimal food arrangement such that we subtract $\texttt{penalty}$ for every person.
+
+Lets have $\texttt{dp}[i][j:\{0,1\}]$ represent the maximum sum of satisfaction of an arrangement of the first $i$ people, given that $j=0/1$ implies whether person $i$ is being used. Let $\texttt{cnt}[i][j]$ represent the number of people used in an optimal arrangement of $\texttt{dp}[i][j]$.  
+
+For our $\texttt{dp}$ transitions, we have
+
+$$
+\{\texttt{dp}[i][0], \texttt{cnt}[i][0]\} = 
+\max\begin{cases} \{\texttt{dp}[i - 1][0], \texttt{cnt}[i - 1][0]\}\\
+\{\texttt{dp}[i - 1][1], \texttt{cnt}[i - 1][1]\}
+\end{cases}
+
+$$
+and 
+$$
+\{\texttt{dp}[i][1], \texttt{cnt}[i][1]\} = 
+\max\begin{cases}\{\texttt{dp}[i - 1][0] + a[i] - \texttt{penalty}, \texttt{cnt}[i - 1][0] + 1\}\\
+\{\texttt{dp}[i - 1][1] + a[i], \texttt{cnt}[i - 1][1]\}\end{cases}
+$$
+
+because we either begin a new subarray or we continue an existing subarray.
+
+Since calculating $\texttt{dp}$ takes $\mathcal O(N)$, and we're binary searching on the range $[0, \sum a[i]]$, the time complexity is $\mathcal{O}(N\log{\sum a[i]})$.
```

Comment:
```suggestion
Since calculating $\texttt{dp}$ takes $\mathcal O(N)$ time, and we're binary searching on the range $[0, \sum a[i]]$, the time complexity is $\mathcal{O}(N\log{\sum a[i]})$.
```